File: Python/problem78.py
maxlim = 10001;
backtrackarr = [];

def amountofoptions(max, sum):
	if (backtrackarr[max][sum] > -1):
		return backtrackarr[max][sum];
	count = 0;
	for i in xrange(1,max+1):
		if (sum == i):
			count += 1;
		elif (sum > i):
			count += amountofoptions(i, sum-i);
		else:
			break;
	backtrackarr[max][sum] = count;
	return count;

def main():
	for i in xrange(maxlim):
		backtrackarr.append([])
		for j in xrange(maxlim):
			backtrackarr[i].append(-1);
	for i in xrange(maxlim):
		a = amountofoptions(i-1,i)+1;
		if (a / 1000000 * 1000000 == a):
			print i
			break;
	print "Done"
	
if __name__ == "__main__":
	main()
	
	

File: problem78.cpp
#include <iostream>
#include "bigint/BigIntegerLibrary.hh"

using namespace std;

const int maxlim = 5001;

BigInteger backtrackarr[maxlim][maxlim];

BigInteger amountofoptions(int max, int sum) {
	if (backtrackarr[max][sum] > -1)
		return backtrackarr[max][sum];
	BigInteger count = 0;
	for (int i = 1; i <= max; i++) {
		if (sum == i) {
			count ++;
		}
		else if (sum > i) {
			count += amountofoptions(i, sum-i);
		}
		else {
			break;
		}
	}
	backtrackarr[max][sum] = count;
	return count;
}

int main() {
	for (int i = 0; i < maxlim; i++) {
		for (int j = 0; j < maxlim; j++) {
			backtrackarr[i][j] = -1;
		}
	}
	for (int i = 1001; i < maxlim; i++) {
		BigInteger a = amountofoptions(i-1,i)+1;
		if (a / 1000000 * 1000000 == a) {
			cout << i;
			break;
		}
	}
	return 0;
}